sklearn/metrics/_pairwise_distances_reduction.pyx

```diff
@@ -0,0 +1,1600 @@
+                            # Guard against eventual -0., causing nan production.
+                            deref(self.neigh_distances)[i][j]
+                            if deref(self.neigh_distances)[i][j] > 0.
+                            else 0
+                        )
+                )
+
+    @final
+    def compute(
+        self,
+        str strategy=None,
+        bint return_distance = False,
+        bint sort_results = False
+    ):
+        """Computes the reduction of vectors (rows) of X on Y.
+
+        Parameters
+        ----------
+        strategy: str, {'auto', 'parallel_on_X', 'parallel_on_Y'}, default=None
+            The chunking strategy defining which dataset parallelization are made on.
+
+            Strategies differs on the dispatching they use for chunks on threads:
+                 - 'parallel_on_X' dispatches chunks of X uniformly on threads.
+                 Each thread then iterates on all the chunks of Y. This strategy is
+                 embarrassingly parallel and comes with no datastructures synchronisation
+                 but is less used in practice (because X is smaller than Y generally).
+                 - 'parallel_on_Y' dispatches chunks of Y uniformly on threads.
+                 Each thread then iterates on all the chunks of X. This strategy is
+                 embarrassingly parallel but uses intermediate datastructures
+                 synchronisation. However it is more useful in practice (because Y is
+                 larger than X generally).
+                 - 'auto' relies on a simple heuristic to choose between
+                 'parallel_on_X' and 'parallel_on_Y'.
+                 - None (default) looks-up in scikit-learn configuration for
+                 `pairwise_dist_parallel_strategy`, and use 'auto' if it is not set.
+
+        return_distance: boolean, default=False
+            Return distances between each X vector and its
+            neighbors if set to True.
+
+        sort_results: boolean, default=False
+            Sort results with respect to distances between each X vector and its
+            neighbors if set to True.
+
+            return_distance must be True if sort_results is set to True.
+
+        Returns
+        -------
+        distances: ndarray of shape (n, k)
+            Distances between each X vector and its neighbors
+            in Y. Only returned if ``return_distance=True``.
+
+        indices: ndarray of shape (n, k)
+            Indices of each neighbor of vectors of X in Y.
+        """
+        if sort_results and not return_distance:
+            raise ValueError("return_distance must be True "
+                             "if sort_results is True.")
+
+        # Temporary datastructures which will be coerced to
+        # numpy arrays on return and then freed.
+        self.neigh_indices = new vector[vector[ITYPE_t]](self.n_X)
+        self.neigh_distances = new vector[vector[DTYPE_t]](self.n_X)
```

Comment:
The lifecycle of `self.neigh_indicies` and `self.neigh_distances` is managed by `compute` which seems weird. If there is an error between here and `free`, there would be a memory leak.

Is it possible to move these into `__init__` and move the `free`s into `__dealloc__`?

Reply:
Definitely, `PairwiseDistancesReduction.compute` can be reworked to alleviate this problem.

I will resolve this with @jeremiedbb's proposals for this method (https://github.com/scikit-learn/scikit-learn/pull/20254#r705114699)

Reply:
This was addressed in 1e2181d.